R/model.R

```diff
@@ -0,0 +1,61 @@
+##' Create a basic `mcstate` model.  Currently nothing here is
+##' validated, and it's likely that users will never actually use this
+##' directly.  Contains data and methods that define a basic model
+##' object, so that we can implement samplers against.  Not all models
+##' will support everything here, and we'll add additional
+##' fields/traits over time to advertise what a model can do.  For
+##' example, models will need to advertise that they are capable of
+##' being differentiated, or that they are stochastic in order to be
+##' used with different methods.
+##'
+##' @title Create basic model
+##'
+##' @param parameters Names of the parameters.  Every parameter is
+##'   named, and for now every parameter is a scalar.  We might relax
+##'   this later to support an `odin`-style structured parameter list,
+##'   but that might just generate a suitable vector of parameter
+##'   names perhaps?  In any case, once we start doing inference it's
+##'   naturally in the R^n, and here n is defined as the length of
+##'   this vector of names.
+##'
+##' @param sample A function to sample from the parameter space.  In
+##'   the case where a model returns a posterior, this is assumed to
+##'   be sampling from the prior.  We'll use this for generating
+##'   initial conditions for MCMC where those are not given, and
+##'   possibly other uses.
```

Comment:
This could be used in situations other than the initial value; it's not yet clear to me where else we'll want this.  Gina talked about sampling from the prior being something that she did as part of model checking so felt like something fairly general

Reply:
I agree with @EmmaLRussell  that this is confusing. I understand the need to sample from the parameter space and not from the model (as it is difficult in general) and happy for the default in Bayesian setting to be sampling from the prior. But if you have a "sample" function in a statistical model, the logic is that it is a way of sampling from this model. Why not calling it "direct_sample" to capture, that it is a simple way of sampling, computationally simpler, but not a sample from the density? it does match well also the concept of sampling from prior distribution in the Bayesian setting

Reply:
Cool, can do.  I think that once we get SMC^2 done we'll have a better sense of the scope of this, and if we only need it for initialising our chains we might diverge anyway